File: scripts/perPlane.py
import ROOT
import os
import pandas as pd
import pyarrow as pa
import pyarrow.parquet as pq
from pathlib import Path

if not hasattr(ROOT, 'sndScifiHit'):
    for library in ["libBase", "libShipData", "libshipLHC", "libsnd_analysis_tools"]:
        ROOT.gSystem.Load(library)




def run_per_plane_extraction(input_files, output_parquet, tree_name="rawConv"):
    chain = ROOT.TChain(tree_name)
    for f in input_files:
        chain.Add(f)

    total_entries = chain.GetEntries()
    print(f"Total entries to process: {total_entries}")

    data = []
    batch_size = 10000
    batch_counter = 0

    for i in range(total_entries):
        if i % 1000 == 0:
            print(f"Processing entry {i}/{total_entries}")

        chain.GetEntry(i)
        event_number = chain.EventHeader.GetEventNumber()
        scifi_hits = chain.Digi_ScifiHits

        planes = {}
        for s in range(1, 6):
            for v in [0, 1]:
                planes[(s, v)] = {'nhits': 0, 'qdc': 0.0}

        for hit in scifi_hits:
            if not hit.isValid(): continue
            station = hit.GetStation()
            is_vertical = 1 if hit.isVertical() else 0

            if (station, is_vertical) in planes:
                planes[(station, is_vertical)]['nhits'] += 1
                planes[(station, is_vertical)]['qdc'] += hit.GetSignal()

        for (station, is_vertical), vals in planes.items():
            if vals['nhits'] > 0:
                data.append({
                    'event_number': event_number,
                    'plane_id': f"S{station}{'V' if is_vertical else 'H'}",
                    'nhits': vals['nhits'],
                    'qdc': vals['qdc']
                })

        if len(data) >= batch_size:
            flush_batch(data, output_parquet, batch_counter)
            data = []
            batch_counter += 1

    if data:
        flush_batch(data, output_parquet, batch_counter)


def flush_batch(data, base_filename, batch_idx):
    df = pd.DataFrame(data)
    table = pa.Table.from_pandas(df)
    path = Path(base_filename)
    filename = f"{path.with_suffix('')}_{batch_idx}.parquet"
    pq.write_table(table, filename, compression='snappy')
    print(f"Saved {filename}")



if __name__ == "__main__":
    import sys
    if len(sys.argv) < 3:
        print("Usage: python perPlane.py <output_parquet> <input_file1> <input_file2> ...")
    else:
        output_name = sys.argv[1]
        inputs = sys.argv[2:]
        run_per_plane_extraction(inputs, output_name)

File: scripts/perPlane.cxx
#include <iostream>
#include <vector>
#include <string>
#include <map>

#include "TFile.h"
#include "TChain.h"
#include "TTree.h"
#include "TClonesArray.h"
#include "TSystem.h"

// SNDLHC headers
#include "sndScifiHit.h"
#include "SNDLHCEventHeader.h"

void perPlane(std::string outputFileName, std::vector<std::string> inputFiles) {
    // Load required libraries
    gSystem->Load("libBase");
    gSystem->Load("libShipData");
    gSystem->Load("libshipLHC");
    gSystem->Load("libsnd_analysis_tools");

    TChain* chain = new TChain("cbmsim");

    // 1. Let TChain handle the wildcard expansion first
    for (const auto& file : inputFiles) {
        chain->Add(file.c_str());
    }

    // 2. Extract the first successfully resolved file to check the tree name
    TObjArray* fileElements = chain->GetListOfFiles();
    if (fileElements->GetEntries() > 0) {
        // TChainElement contains the actual resolved file path
        TChainElement* element = (TChainElement*)fileElements->At(0);
        TFile* firstFile = TFile::Open(element->GetTitle());
        if (firstFile) {
            if (!firstFile->Get("cbmsim")) {
                chain->SetName("rawConv");
            }
            firstFile->Close();
        }
    } else {
        std::cerr << "Error: No files found matching the input." << std::endl;
        return;
    }

    long nEntries = chain->GetEntries();
    std::cout << "Processing " << nEntries << " entries from tree: " << chain->GetName() << std::endl;

    TClonesArray* scifiHits = nullptr;
    SNDLHCEventHeader* eventHeader = nullptr;
    chain->SetBranchAddress("Digi_ScifiHits", &scifiHits);

    bool isData = (std::string(chain->GetName()) == "rawConv");
    if (isData) {
        chain->SetBranchAddress("EventHeader", &eventHeader);
    }

    // Set up output file and tree
    TFile* outFile = new TFile(outputFileName.c_str(), "RECREATE");
    TTree* outTree = new TTree("perPlane", "SciFi Per Plane Data");

    int out_event;
    int out_plane_id;
    int out_nhits;
    double out_qdc;

    outTree->Branch("event_number", &out_event);
    outTree->Branch("plane_id", &out_plane_id);
    outTree->Branch("nhits", &out_nhits);
    outTree->Branch("qdc", &out_qdc);

    for (long i = 0; i < nEntries; ++i) {
        chain->GetEntry(i);
        if (i % 10000 == 0) std::cout << "Event " << i << " / " << nEntries << std::endl;

        out_event = eventHeader->GetEventNumber();

        // plane_id -> {nhits, qdc}
        // plane_id formula: station * 10 + isVertical
        std::map<int, std::pair<int, double>> planes;

        int nHits = scifiHits->GetEntries();
        for (int j = 0; j < nHits; ++j) {
            sndScifiHit* hit = (sndScifiHit*)scifiHits->At(j);
            if (!hit->isValid()) continue;

            int plane_id = hit->GetStation() * 10 + (hit->isVertical() ? 1 : 0);
            planes[plane_id].first++;
            planes[plane_id].second += hit->GetSignal();
        }

        for (auto const& [id, data] : planes) {
            out_plane_id = id;
            out_nhits = data.first;
            out_qdc = data.second;
            outTree->Fill();
        }
    }

    outTree->Write();
    outFile->Close();
    std::cout << "Done. Output saved to " << outputFileName << std::endl;
}

// Entry point for running via: root -l -b -q 'perPlane.cxx("out.root", {"in1.root", "in2.root"})'
void perPlane() {
    std::cout << "Usage: root -l 'perPlane.cxx(\"output.root\", {\"input.root\"})'" << std::endl;
}
